compiler/rustc_middle/src/ty/mod.rs

```diff
@@ -522,6 +522,9 @@ impl<'tcx> Predicate<'tcx> {
     #[instrument(level = "debug", skip(tcx), ret)]
     pub fn is_coinductive(self, tcx: TyCtxt<'tcx>) -> bool {
         match self.kind().skip_binder() {
+            // Always coinductive in the new solver.
+            _ if tcx.trait_solver_next() => true,
```

Comment:
See https://hackmd.io/frHMYxD5S4-rWlxkx_eWNw#2-Coinduction-is-necessary-for-this-to-work